packages/react-router/lib/router/define-route.ts

```diff
@@ -66,7 +64,7 @@ type Params<Param extends string> = Pretty<
   & {[K in Param as K extends `${infer P}?` ? P : never]?: string}
 >
 
-type IsDefined<T> = undefined extends T ? false : true;
+type IsDefined<T> = Equal<T, undefined> extends true ? false : true;
```

Comment:
Before we checked if `undefined` extended `T`, which is true for unions containing `undefined`. Now we check if `T` is _exactly_ `undefined`. This works since we now default all of the data types in `defineRoute` and `defineRootRoute` generics to be `undefined` (`Data | undefined = undefined`)